experimental/grpc_attestation/src/attestation.rs

```diff
@@ -0,0 +1,210 @@
+                Ok(attestation_request)
+            } else {
+                Err(anyhow!("Received incorrect message type"))
+            }?;
+        info!(
+            "Server received attestation request: {:?}",
+            attestation_request
+        );
+
+        // Generate session key.
+        let key_negotiator =
+            DiffieHellmanNegotiator::create().expect("Couldn't create key negotiator");
+        let public_key = key_negotiator
+            .public_key()
+            .expect("Couldn't get public key");
+        let session_key = key_negotiator
+            .derive_session_key(&attestation_request.public_key)
+            .expect("Couldn't derive session key");
+
+        let encryptor = AeadEncryptor::new(&session_key);
+
+        // Generate attestation info with a TEE report.
+        // TEE report contains a hash of the server's public key.
+        let public_key_hash = get_sha256(&public_key);
+        let tee_report = Report::new(&public_key_hash);
+        let attestation_info = AttestationInfo {
+            report: tee_report,
+            certificate: tee_certificate,
+        };
+        let serialized_attestation_info = attestation_info
+            .to_string()
+            .expect("Couldn't serialize attestation info");
+
+        let attestation_response = GetAttestedDataResponse {
+            response_type: Some(ResponseType::AttestationResponse(AttestationResponse {
+                public_key,
+                attestation_info: serialized_attestation_info.as_bytes().to_vec(),
+            })),
+        };
+
+        Ok((attestation_response, encryptor))
+    }
+}
+
+#[tonic::async_trait]
+impl GrpcAttestation for AttestationServer {
+    type GetAttestedDataStream = Pin<
+        Box<dyn Stream<Item = Result<GetAttestedDataResponse, Status>> + Send + Sync + 'static>,
+    >;
+
+    async fn get_attested_data(
+        &self,
+        request_stream: Request<tonic::Streaming<GetAttestedDataRequest>>,
+    ) -> Result<Response<Self::GetAttestedDataStream>, Status> {
+        let tee_certificate = self.tee_certificate.clone();
+        let request_handler = self.request_handler;
+        info!("Server received request stream");
+
+        let request_stream = request_stream.into_inner();
+        let response_stream = async_stream::try_stream! {
+            let mut receiver = Receiver { request_stream };
+            let (attestation_response, encryptor) = Self::attest(&mut receiver, tee_certificate)
+                .await
+                .expect("Couldn't attest to the client");
+            yield attestation_response;
+
+            let mut handler = RequestHandler::new(encryptor, request_handler);
+            while let Some(response) = handler.handle_request(&mut receiver).await.expect("Couldn't process stream") {
+                yield response
+            }
+        };
+
+        Ok(Response::new(Box::pin(response_stream)))
+    }
```

Comment:
Is it possible to encapsulate only the stream-related logic into something that makes it look essentially like implementing a function with the following type:

```rust
async fn handle(request_stream: Stream<...>, response_stream: Sink<...>) { ... }
```

(using https://docs.rs/futures/0.3.15/futures/stream/trait.StreamExt.html)


or even something with  https://docs.rs/tokio/1.6.0/tokio/sync/mpsc/index.html (though probably not worth the complexity with the buffering) :

```rust
async fn handle(request_stream: Receiver<...>, response_stream: Sender<...>) { ... }
```